assembly/src/sema/context.rs

```diff
@@ -0,0 +1,217 @@
+use crate::{
+    ast::*,
+    diagnostics::{Diagnostic, Report, Severity, SourceFile},
+    Felt, LibraryPath, Span, Spanned,
+};
+use alloc::{
+    boxed::Box,
+    collections::{BTreeMap, BTreeSet},
+    sync::Arc,
+    vec::Vec,
+};
+
+use super::{SemanticAnalysisError, SyntaxError, SyntaxWarning};
+
+/// This maintains the state for semantic analysis of a single [Module].
+pub struct AnalysisContext {
+    source_code: Arc<SourceFile>,
+    module: Option<Box<Module>>,
```

Comment:
This is a bit convoluted primarily due to Rust's borrowing semantics. During analysis, there are times where we need to be able to hold a mutable reference both to a `Module` and to a procedure that belongs to that module. In order to do so, after we have initially constructed a `Module` from the parsed `Form`s, we take the procedures out of the `Module`, and then run our rewrites/analysis on each procedure as a separate object. This allows us to use the mutable APIs of `Module` while holding a mutable reference to a procedure that belongs to that `Module.

There isn't a great way to refactor this, and this kind of thing is inherently messy anyway, but here's how I'd modify this:

1. Remove the `define_import` and `define_procedure` helpers from `AnalysisContext`, and make them free functions in `sema/mod.rs`
2. Remove the `module` field from `AnalysisContext` entirely, and pass it around on the stack instead.
3. Keep the other state tracking/error management in `AnalysisContext`

I think this is slightly less convoluted than the current code, albeit not by much. It does have the benefit of removing the pseudo-optionality of the `module` field though, which I think is likely to be the biggest source of confusion.

Thoughts?

Reply:
If I'm understanding the code correctly, the only reason why we need to have `module` in `AnalysisContext` is so that we can call `define_import()` and `define_procedure()` on it. What if we keep track of imports and procedures in `AnalysisContext` separately? Then, after the analysis context is complete, we could build a `Module` from the data contained in it.

This way, `AnalysisContext` becomes a kind of a "builder" for `Module` and module could become a pure read-only struct (e.g., no need to have mutable methods on it).